### PR TITLE
Implement getSupportInfo() and use it for inference

### LINKED ISSUE #3023: Expose list of supported parsers, etc.
Currently every time Prettier adds another language, a lot of editor integrations have to release another version of their plugin to support it. This might be adding more language types, extensions, etc.

I'm wondering if we'd be able to support some kind of feature detection within Prettier that editor integrations can call and run more dynamically. Something along the lines of:

```js
require('prettier').getSupportInfo();
```

Which might return:

```js
{
  languages: [
    {
      language: "javascript",
      parsers: ["babylon", "flow"],
      ids: ["javascript", "javascriptreact"],
      extensions: [".js", ".jsx", ".mjs"]
    },
    {
      language: "typescript",
      parsers: ["typescript"],
      ids: ["typescript", "typescriptreact"],
      extensions: [".ts", ".tsx"]
    },
    ...
  ]
}
```

Would be nice to release this alongside Markdown support.

Love to get some feedback from some plugin maintainers on whether this kind of thing could be useful.

cc. @robwise @CiGit @mitermayer

### LINKED ISSUE #3058: The experimental HTML support unexpectedly leaks through the CLI
Edit by @lydell:

```
prettier foo.html
```

The above formats foo.html as HTML. It should try to format it as JS (as usual with unknown file extensions, until  #2884 is resolved).

Original issue description:

`<script src="/build.js"></script>`

is being prettified to

`<script src="/build.js" />`

which prevents the script from running for some reason.

### PR DESCRIPTION
**Pull request by azz**
TODO

 - [x] CLI
 - [x] Docs

This PR adds a new function to the prettier API: `getSupportInfo()`.

It returns an object, currently with a `languages` property, which is an array of:

```js
{
  name: "SCSS", // display name for the language
  since: "1.4.0", // version it was added to prettier
  parsers: ["scss"], // built-in parsers that support this language
  group: "CSS", // metadata used to group related languages
  tmScope: "source.scss", // TextMate scope, used by prettier-atom
  aceMode: "scss",
  codemirrorMode: "css",
  codemirrorMimeType: "text/x-scss",
  extensions: [".scss"], // extensions used in parser inference
  files: [], // filenames without extensions, also used in parser inference
  liguistLanguageId: 329, // GitHub linguist language ID
  vscodeLanguageIds: ["scss"] // used by prettier-vscode
},
```

`getSupportInfo` accepts an optional argument, `version`, e.g. `prettier.getSupportInfo("1.4.0")`, this will filter the list to only include languages that version 1.4.0 of prettier support. This includes changing `less/scss/css` parsers back to `postcss`.

In addition, this table is now used by Prettier internally to perform parser inference based on filename and extension.

Maybe we should also add a CLI version?

Closes #3023 
Fixes #3058

### CODE REVIEW

File: src/support.js [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+function getSupportInfo(version) {
+  if (!version) {
+    version = currentVersion;
+  }
+
+  const usePostCssParser = semver.lt(version, "1.7.1");
+
+  const languages = supportTable
+    .filter(language => {
+      return language.since && semver.gte(version, language.since);
```
Comment (ikatyang) [inline src/support.js]: Can be simplified to `language => language...`

File: tests_integration/__tests__/support-info.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+describe("API getSupportInfo()", () => {
+  test("no arguments", () => {
+    expect(prettier.getSupportInfo()).toMatchSnapshot();
+  });
+
+  const testVersions = ["0.0.0", "1.0.0", "1.4.0", "1.5.0", "1.7.1", "1.8.0"];
+
+  testVersions.forEach(version => {
+    test(`with version ${version}`, () => {
+      expect(prettier.getSupportInfo(version)).toMatchSnapshot();
```
Comment (ikatyang) [inline tests_integration/__tests__/support-info.js]: Can we only snapshot the `{ [language]: parsers }` information? It's hard to read its snapshot and other information seems unrelated here.

File: src/support.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
+"use strict";
+
+const semver = require("semver");
+const currentVersion = require("../package").version;
+
+// Based on:
+// https://github.com/github/linguist/blob/master/lib/linguist/languages.yml
+
+const supportTable = [
```
Comment (ikatyang) [inline src/support.js]: Maybe we should add a script to generate this table from `linguist` so that it'd be easy to update in the future.

**Comment by ikatyang** [comment]
> Maybe we should also add a CLI version?

Yeah, the CLI seems necessary for non-JS editor, e.g. vim-prettier.

File: package.json [hunk 4]
```diff
@@ -42,6 +42,7 @@
     "postcss-values-parser": "1.3.1",
     "remark-frontmatter": "1.1.0",
     "remark-parse": "4.0.0",
+    "semver": "^5.4.1",
```
[the remark this answers is not in the document]
Comment (azz) [reply package.json]: Has `save-prefix ""` been removed from `yarn`?

[the remark this answers is not in the document]
Comment (ikatyang) [reply package.json]: No, it works for me (v1.2.1).

File: src/support.js (hunk 3 above)
Comment (azz) [reply src/support.js]: Maybe. I ended up changing quite a few things though, added more extensions, filenames, added groups, etc.

File: src/support.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+      ".workbook"
+    ],
+    filenames: ["README"],
+    tmScope: "source.gfm",
+    linguistLanguageId: 222,
+    vscodeLanguageIds: ["markdown"]
+  },
+  {
+    name: "HTML",
+    since: undefined, // unreleased
```
Comment (suchipi) [inline src/support.js]: No plans to release this yet (in eg 1.8.0), right?

Comment (azz) [reply src/support.js]: Probably not. Depends on #2976

**Comment by mitermayer** [comment]
Are there any plans on having this function exposed through the CLI ? I believe the info table supplied by the CLI does not  need to be so verbose so would probably be a much simpler output than the one returned by `getSupportInfo()`

**Comment by azz** [comment]
What kind of output would you expect from the CLI version?

**Comment by azz** [comment]
Ping @mitermayer, would JSON output or a parseable table be good for CLI?

**Comment by mitermayer** [comment]
Hi @azz,

Vim has a function `json_decode()` however to better leverage that we need to make sure that the output of `getSupportInfo()` would be a plain json with nothing else, otherwise we would still need to use regex to extract the JSON node, and in that case a table would make more sense.

This is the documentation info on json_decode() function for vim 8:

```
json_decode({string})					*json_decode()*
		This parses a JSON formatted string and returns the equivalent
		in Vim values.  See |json_encode()| for the relation between
		JSON and Vim values.
		The decoding is permissive:
		- A trailing comma in an array and object is ignored.
		- More floating point numbers are recognized, e.g. "1." for
		  "1.0".
		However, a duplicate key in an object is not allowed. *E938*
		The result must be a valid Vim type:
		- An empty object member name is not allowed.
		- Duplicate object member names are not allowed.
```

**Comment by mitermayer** [comment]
I actually am not sure we can leverage from the `json_decode` to parse arbitrary json. So I would suggest printing as a table since it would also be nicer for CLI users, and on the vim side we can use regex to extract info output

**Comment by azz** [comment]
Let's just do 

```js
console.log(
  prettier.format(
    JSON.stringify(prettier.getSupportInfo()),
    { parser: "json" }
  )
);
```

?

**Comment by mitermayer** [comment]
We can always start with that, then i can try to to add that to vim-prettier and see how that goes. Can always submit another PR later to allow the getSupportInfo() cli command to accept other output formats other than JSON if required

**Comment by azz** [comment]
@mitermayer do you want to give this a try with json_decode before we merge?

**Comment by mitermayer** [comment]
I won't be able to test this today, so feel free to merge it as it is! Can always wait for next release in case this is not sufficient.
